keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,309 @@
+
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: Bucket
+metadata:
+  name:
+  namespace:
+  labels:
+    cosi.io/provisioner: [1]
+  finalizers:
+  - cosi.io/finalizer [2]
+spec:
+  protocol: [3]
+  bucketPrefix: [4]
+  bucketClassName: [5]
+  accessSecretName: [6]
+status:
+  bucketContentName: [7]
+  phase: [8]
+  conditions: 
+```
+1. `labels`: should be added by controller.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+1. `finalizers`: should be added by controller to defer `Bucket` deletion until backend deletion ops succeed.
+1. `protocol`: specifies the desired protocol.  One of {“s3”, “gcs”, or “azureBlob”}.
+1. `bucketPrefix`: (Optional) prefix prepended to a randomly generated bucket name, eg. "YosemitePhotos-". If empty no prefix is appended.
+1. `bucketClassName`: Name of the target `BucketClass`.
+1. `userSecretName`: (optional) Secret in the Bucket’s namespace storing credentials to be used by a workload for bucket access.
+1. `bucketContentName`: Name of a bound `BucketContent`.
+1. `phase`: 
+   - *Pending*: The controller has detected the new `Bucket` and begun provisioning operations
+   - *Bound*: Provisioning operations have completed and the `Bucket` has been bound to a `BucketContent`.
+
+#### BucketContent
+
+A cluster-scoped resource representing an object store bucket. The `BucketContent` is expected to store stateful data relevant to bucket deprovisioning. The `BucketContent` is bound to the `Bucket` in a 1:1 mapping. For MVP a `BucketContent` is not reused.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketContent
+Metadata:
+  name: [1]
+  labels:
+    cosi.io/provisioner: [2]
+  finalizers:
+  - cosi.io/finalizer [3]
+spec:
+  provisioner: [4]
+  releasePolicy: [5]
+  accessMode: [6]
+  bucketClassName: [7]
+  bucketRef: [8]
+    name:
+    namespace:
+    uuid:
+  secretRef: [9]
+    name:
+    namespace:
+  protocol: [10]
+    type: ""
+    azureBlob: [11]
+      containerName:
+    s3: [12]
+      endpoint:
+      bucketName:
+      region:
+      signatureVersion:
+    gcs: [13]
+      bucketName:
+      privateKeyName:
+      projectId:
+      serviceAccount:
+  parameters: [14]
+status:
+  message: [15]
+  phase: [16]
+  conditions:
+```
+1. `name`: Generated in the pattern of `<BUCKET-CLASS-NAME>'-'<RANDOM-SUFFIX>`. 
+1. `labels`: should be added by controller..  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+1. `finalizers`: should be added by controller to defer `Bucket` deletion until backend deletion ops succeed.
```

Comment:
If the `Bucket` can't be deleted before the `BucketContent` is, then is the `secretRef` here necessary, since this could be retrieved from the matching `Bucket`?

Reply:
@NicolasT (Note: in the latest version of the KEP the API name `BucketContent` is replaced by `Bucket`)
The finalizers allow COSI to orchestrate deletion. They typical trigger is the user deletes their `BucketRequest`; however, an admin could delete a `Bucket` and we need to account for that. <-- @wlan0 ?